### DISCUSSION TITLE
Get `stmt.expaned_sql()` if `stmt.execute(...)` fails

### DISCUSSION THREAD

**Discussion by maxwell8888** [rusqlite/rusqlite#1185] [Questions] [answered]
In the below example, say the binding of parameters is successful, but the SQL fails because of a foreign key constraint, for example. Is it possible to get the bound sql statement, which is useful for debugging why the SQL failed?  I used `.raw_bind_parameter()` as a workaround which seems fine but has the following statement in the docs which made me think maybe I shouldn't be using it and there could be a better way.

> This should not generally be used, but is available for special cases such as:
binding parameters where a gap exists.
binding named and positional parameters in the same query.
separating parameter binding from query execution.

```rust
let mut stmt = conn.prepare("INSERT INTO test (foo, var) VALUES (?, ?)")?;
stmt.execute(params![1i32, 2i32])?;
```

**Comment by gwenn** [comment, marked as answer]
```toml
rusqlite = { version="0.27", features = ["bundled"] }
```
```rs
use rusqlite::*;
use rusqlite::types::Null;

fn main() -> Result<()> {
    let conn = Connection::open_in_memory()?;
    conn.execute("CREATE TABLE test (foo INT NOT NULL, var INT NOT NULL);",  [])?;
    let mut stmt = conn.prepare("INSERT INTO test (foo, var) VALUES (?, ?)")?;
    if let Err(_) = stmt.execute(params![Some(1i32), Null]) {
        println!("{:?}", stmt.expanded_sql());
    }
    Ok(())
}
```
```sh
% cargo run
...
Some("INSERT INTO test (foo, var) VALUES (1, NULL)")
```

**Reply by maxwell8888** [reply]
Ah I feel stupid I didn't know I could just do that! Thank you!
